File: leetcode/src/zigzaggame.cpp
/*
 * the string "paypalistring" is written in a zigzag pattern on a given number of rows like this:
 * p   a   t   g
 * a p l s r n
 * y   i   i 
 * */

#include "stdio.h"
#include <iostream>
#include <string>
#include <cstring>
#include "stdlib.h"
#include <vector>
using namespace std;

string zigzag(const string& str, int rows){
    if(rows==0)        return string();
    if(rows==1)        return string(str);
    int n = str.size();
    vector<char> pvec[rows];
    bool up=false;
    for(int i=0, r=0; i<n;i++){
        pvec[r].push_back(str[i]);
        if(up)  r--;
        else    r++;
        
        if(r==-1){
            r = 1;
            up = false;
        }
        if(r==rows){
            r = rows-2;
            up = true;
        }
    }
    string res;
    for(int i=0;i<rows;i++){    //take care of the string build and output
        int l = pvec[i].size();
        char cstr[l+1];
        memset(cstr, 0, sizeof(char)*(l+1));
        for(int j=0;j<l;j++){
            cstr[j] = pvec[i][j];
        }
        res.append(string(cstr));
    }
    return res;
}

int main(int, char**){
    string str;
    while(1){
        cout<<"please input rows:"<<endl;
        if(getline(cin, str)==0 || str.empty())    
          break;
        int rows = atoi(str.c_str());
        
        cout<<"please input string:"<<endl;
        if(getline(cin, str)==0 || str.empty())    
          break;
        printf("ZigZag game: %s\n", zigzag(str, rows).c_str());
    }
    return 0;
}

File: leetcode/java/zigzagGame.java
/*
 * the string "paypalishiring" is written in a zigzag pattern on a given number of rows like this: 
 * p     i     n
 * a   l s   i g
 * y a   h r
 * p     i
 * */
import java.io.*;
import java.util.*;

public class zigzagGame{
    public String convert(String s, int nRows){
        if(nRows==0)    return new String();
        if(nRows==1)    return new String(s);
        StringBuilder pBuilders[] = new StringBuilder[nRows];    //StringBuilder is perfect container for insert/append char
        for(int r=0;r<nRows;r++){
            pBuilders[r] = new StringBuilder();
        }
        boolean up = false;
        for(int i=0, r=0;i<s.length();i++){
            pBuilders[r].append(s.charAt(i));
            if(up)      r--;
            else        r++;
            
            if(r==-1){
                r = 1;
                up = false;
            }
            if(r==nRows){
                r = nRows-2;
                up = true;
            }
        }
        StringBuilder res = new StringBuilder();
        for(int r=0;r<nRows;r++){
            res.append(pBuilders[r].toString());
        }
        return res.toString();
    }


    public static void main(String[] args){
        Scanner scan  = new Scanner(System.in);
        zigzagGame zg = new zigzagGame();
        while(true){
            System.out.println("please input string:");
            String str = scan.nextLine();
            if(str.isEmpty())    break;
            
            System.out.println("please input nRows:");
            String sRows = scan.nextLine();
            if(sRows.isEmpty())    break;
            int nRows = new Integer(sRows);
            
            System.out.println(zg.convert(str, nRows));            
        }
    }
}
